packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,269 @@
+
+        // Map amp tags to their required extension(s)
+        if (tagName.startsWith('amp-')) {
+          tagToExtensionsMapping.set(
+              tagName,
+              tag.requiresExtension,
+          );
+        }
+        // Hack: fix missing attribute dependencies (e.g. amp-img => lightbox => amp-lightbox-gallery)
+        tag.attrs.filter((a) => manualAttributeToExtensionMapping.has(a.name))
+            .forEach((a) => {
+              a.requiresExtension = [manualAttributeToExtensionMapping.get(a.name)];
+            });
+
+        // Map attributes to tags and extensions (e.g. amp-img => amp-fx => amp-fx-collection)
+        tag.attrs.filter((a) => a.requiresExtension && a.requiresExtension.length > 0)
+            .forEach((a) => {
+              let attributeMapping = tagToAttributeMapping.get(tagName);
+              if (!attributeMapping) {
+                attributeMapping = [];
+                tagToAttributeMapping.set(tagName, attributeMapping);
+              }
+              attributeMapping.push(a);
+            });
+
+        // Map tags to bindable attributes which are named link `[text]`
+        const bindableAttributes = tag.attrs.filter((a) => a.name.startsWith('['))
+            .map((a) => a.name.substring(1, a.name.length - 1));
+        tagToBindAttributeMapping.set(tagName, new Set(bindableAttributes));
+      }
+      return {
+        extensionsMap,
+        tagToExtensionsMapping,
+        tagToAttributeMapping,
+        tagToBindAttributeMapping,
+      };
+    });
+  }
+
+  async transform(tree, params) {
+    if (!this.enabled) {
+      return;
+    }
+    const html = tree.root.firstChildByTag('html');
+    const head = html.firstChildByTag('head');
+    if (!head) return;
+
+    // Extensions which need to be imported
+    const extensionsToImport = new Set();
+    // Keep track of existing extensions imports to avoid duplicates
+    const existingImports = new Set();
+
+    // Some AMP components need to be detected in the head (e.g. amp-access)
+    await this.findExtensionsToImportInHead_(head, extensionsToImport, existingImports);
+
+    // Most AMP components can be detected in the body
+    const body = html.firstChildByTag('body');
+    await this.findExtensionsToImportInBody_(body, extensionsToImport);
+
+    if (extensionsToImport.length === 0) {
+      // Nothing to do
+      return;
+    }
+
+    // We use this for adding new import elements to the header
+    const referenceNode = findMetaViewport(head);
+
+    // Support custom runtime URLs
+    const host = calculateHost(params);
+    for (const extensionName of extensionsToImport) {
+      if (!existingImports.has(extensionName)) {
+        const extension = (await this.extensionSpec).extensionsMap.get(extensionName.trim());
+        this.log_.debug('auto importing', extensionName);
+        // Use the latest version by default
+        const version = extension.version[extension.version.length - 1];
```

Comment:
This makes me feel a bit uneasy, are we 100% sure the latest is the last in the array?